xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    /**
+     * Sends a DiscoveryRequest with the given information as an NACK. NACK takes the previous
+     * accepted version.
+     */
+    private void sendNackRequest(String typeUrl, Collection<String> resourceNames, String nonce,
+        String message) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      String versionInfo = "";
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        versionInfo = ldsVersion;
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        versionInfo = rdsVersion;
+      }
+      // TODO(chengyuanzhang): cases for CDS/EDS.
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node)
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .setErrorDetail(com.google.rpc.Status.newBuilder().setMessage(message))
+              .build();
+      requestWriter.onNext(request);
+    }
+  }
+
+  /**
+   * Returns {@code true} iff {@code hostName} matches the domain name {@code pattern} with
+   * case-insensitive.
+   *
+   * <p>Wildcard pattern rules:
+   * <ol>
+   * <li>A single asterisk (*) matches any domain.</li>
+   * <li>Asterisk (*) is only permitted in the left-most or the right-most part of the pattern,
+   *     but not both.</li>
+   * </ol>
+   */
+  @VisibleForTesting
+  static boolean matchHostName(String hostName, String pattern) {
+    // Basic sanity checks
+    if (hostName.length() == 0 || hostName.startsWith(".") || hostName.endsWith(".")) {
+      // Invalid host name
+      return false;
+    }
+    if (pattern.length() == 0 || pattern.startsWith(".") || pattern.endsWith(".")) {
+      // Invalid pattern/domain name
+      return false;
+    }
+
+    hostName = hostName.toLowerCase(Locale.US);
+    pattern = pattern.toLowerCase(Locale.US);
+    // hostName and pattern are now in lower case -- domain names are case-insensitive.
+
+    if (!pattern.contains("*")) {
+      // Not a wildcard pattern -- hostName and pattern must match exactly.
+      return hostName.equals(pattern);
+    }
+    // Wildcard pattern
+
+    if (pattern.length() == 1) {
+      return true;
+    }
+
+    int index = pattern.indexOf('*');
+
+    // At most one asterisk (*) is allowed.
+    if (pattern.indexOf('*', index + 1) != -1) {
+      return false;
+    }
+
+    // Asterisk can only match prefix or suffix.
+    if (index != 0 && index != pattern.length() - 1) {
+      return false;
+    }
+
+    // Optimization: check whether hostName is too short to match the pattern. hostName must be at
+    // least as long as the pattern because asterisk has to match one or more characters.
```

Comment:
No much performance benefit to optimize here. Remove it or move it before `hostName.toLowerCase`

Reply:
Fine, removed.